base/compiler/tfuncs.jl

```diff
@@ -460,7 +460,25 @@ end
 add_tfunc(Core._typevar, 3, 3, typevar_tfunc, 100)
 add_tfunc(applicable, 1, INT_INF, (@nospecialize(f), args...)->Bool, 100)
 add_tfunc(Core.Intrinsics.arraylen, 1, 1, @nospecialize(x)->Int, 4)
-add_tfunc(arraysize, 2, 2, (@nospecialize(a), @nospecialize(d))->Int, 4)
+
+function arraysize_tfunc(@nospecialize(arytype), @nospecialize(dim))
+    hasintersect(widenconst(arytype), Array) || return Bottom
+    hasintersect(widenconst(dim), Int) || return Bottom
+    return Int
+end
+add_tfunc(arraysize, 2, 2, arraysize_tfunc, 4)
+
+function arraysize_nothrow(argtypes::Vector{Any})
+    length(argtypes) == 2 || return false
+    ary = argtypes[1]
+    dim = argtypes[2]
+    ary ⊑ Array || return false
+    if isa(dim, Const)
+        dimval = dim.val
+        return isa(dimval, Int) && dimval > 0
```

Comment:
Does an OOMError count? Probably not.

Reply:
I don't think `arraysize` can yield OOMError though? AFAIU array resizing may result in OOMError, but all them are handled `foreigncall`, which currently is handled conservatively i.e. considered as "may throw" always.

Reply:
Yes, I read a bit too quickly and thought it was about array construction. Might have mixed it up with some other PR.

Reply:
yeah array constructions are handled by #43565, and it does account for OOMError.